packages/tailwindcss/src/variants.ts

```diff
@@ -711,6 +711,35 @@ export function createVariants(theme: Theme): Variants {
 
   staticVariant('inert', ['&:is([inert], [inert] *)'])
 
+  variants.compound('in', Compounds.StyleRules, (ruleNode, variant) => {
+    if (variant.modifier) return null
+
+    let didApply = false
+
+    walk([ruleNode], (node, { path }) => {
+      if (node.kind !== 'rule') return WalkAction.Continue
+
+      // Throw out any candidates with variants using nested style rules
+      for (let parent of path.slice(0, -1)) {
+        if (parent.kind !== 'rule') continue
+
+        didApply = false
+        return WalkAction.Stop
+      }
+
+      // Replace `&` in target variant with `*`, so variants like `&:hover`
+      // become `:where(*:hover) &`. The `*` will often be optimized away.
+      node.selector = `:where(${node.selector.replaceAll('&', '*')}) &`
```

Comment:
This is basically the same implementation as the `has-*` variant, but the selector is a bit different.